core/src/main/scala/org/apache/spark/sql/execution/columnar/encoding/DictionaryEncoding.scala

```diff
@@ -106,96 +118,46 @@ abstract class DictionaryDecoderBase
       case _ => throw new UnsupportedOperationException(
         s"DictionaryEncoding not supported for $dataType")
     }
-    baseCursor = position
     position
   }
 
-  private[sql] def initializeBeforeFinish(dictionaryBytes: AnyRef,
-      dictionaryCursor: Long, dictionaryLen: Int, cursor: Long, dataType: DataType): Unit = {
-    initializeDictionary(dictionaryBytes, dictionaryCursor,
-      dictionaryLen, dictionaryLen, dataType)
-    // base cursor will start at the given cursor since dictionary is separate
-    baseCursor = cursor
-  }
-
-  override def nextUTF8String(columnBytes: AnyRef, cursor: Long): Long =
-    cursor + 2
-
-  override def absoluteUTF8String(columnBytes: AnyRef, position: Int): Long = {
-    // TODO: PERF: optimize for local index access case by filling
-    // in the dictionary array lazily on access
-    baseCursor + ((position - numNullsUntilPosition(columnBytes, position)) << 1)
-  }
-
-  override def readUTF8String(columnBytes: AnyRef, cursor: Long): UTF8String =
-    stringDictionary.getString(columnBytes, ColumnEncoding.readShort(columnBytes, cursor))
+  // TODO: PERF: optimize for local index access case by filling
+  // in the dictionary array lazily on access
+  override def readUTF8String(columnBytes: AnyRef, nonNullPosition: Int): UTF8String =
+  stringDictionary.getString(columnBytes,
```

Comment:
@hbhanawat How it does matter, if we create string at scan time or on a upper operator like join. The cost should be same as I understood. However in this approach if we can push some filters( filter on string columns can be looked up on a dictionary and discard the batch altogether)  to dictionary then probably we don’t have to create a lot of objects.